prometheus_juju_exporter/collector.py

```diff
@@ -137,14 +129,19 @@ async def _get_machine_stats(
         :param str model_name: the name of the model the machines are in
         :param str gauge_name: the name of the gauge
         """
+        virt_mac_prefixes = self.config["machine"]["virt_macs"].get(str).split(",")
+
         for machine in machines.values():
             # check machine type with the prefix of its mac address
             interfaces = machine["network-interfaces"].values()
             machine_type = MachineType.METAL
 
             for i in interfaces:
```

Comment:
this is largely a matter of stylistic choice, so it's really up to you. I would personally rewrite this bit to make it look a bit closer to natural language. Something like

```python
for machine in machines.values():
    mac_addresses = [_["mac-address"] for _ in machine["network-interfaces"].values()]
    machine_type = MachineType.KVM if any(mac.startswith(virt_mac_prefixes) for mac in mac_addresses) else MachineType.METAL
```

Reply:
My initial implementation was similar to your suggestion ([collector.py#L127](https://github.com/agileshaw/prometheus-juju-exporter/blob/89ed6763745410ae4169776b844b69c9491400fd/prometheus_juju_exporter/collector.py#L127)). However, I had a [discussion](https://github.com/canonical/prometheus-juju-exporter/pull/4#discussion_r1047124654) with Eric on this and the "for-loop + break" approach seems more efficient since we don't need to go over all interfaces.

Reply:
that's true, however:
* this is not meant to be high performance code, I think we should favor readability and maintainability (which can be quite a subjective metric :))
* if we really were concerned with performance I'd wager we'd probably gain more by caching values than micro-optimizing small loops

Reply:
I think that more suitable will be to use separate function here, function `refresh_cache` is quite long and there are multiple operations.
What do you think about using something like this:
```python
def _get_machine_type(machine, virt_mac_prefixes):
  interfaces = machine.get("network-interfaces", {}).values()
  for interface in interfaces:
    if interface["mac-address"].startswith(tuple(virt_mac_prefixes)):
       return MachineType.KVM

  return MachineType.METAL
```
Like this we will have efficient code and easy to maintenance, or at least in my point of view it will be simple to test and maintenance.

Reply:
`refresh_cache` doesn't contain machine type detection code. I think you meant `_get_machine_stats`. But I agree with you, it would be better to have it in a separate function. Code updated.